lib/mlir/Dialect/Clift/IR/ImportModel.cpp

```diff
@@ -0,0 +1,599 @@
+
+    for (const model::UnionField &Field : ModelType.Fields()) {
+      auto const FieldType = rc_recur
+        getQualifiedType(Field.Type(), /*RequireComplete=*/true);
+      if (not FieldType)
+        rc_return nullptr;
+      auto const Attribute = make<clift::FieldAttr>(0,
+                                                    FieldType,
+                                                    Field.CustomName());
+      if (not Attribute)
+        rc_return nullptr;
+      Fields.push_back(Attribute);
+    }
+
+    rc_return make<clift::UnionType>(ModelType.ID(),
+                                     ModelType.OriginalName(),
+                                     Fields);
+  }
+
+  RecursiveCoroutine<clift::TypeDefinition>
+  getTypeAttribute(const model::Type &ModelType, bool &RequireComplete) {
+    switch (ModelType.Kind()) {
+    case model::TypeKind::CABIFunctionType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::CABIFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::EnumType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::EnumType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::RawFunctionType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::RawFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::StructType: {
+      auto const &T = llvm::cast<model::StructType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::TypedefType: {
+      auto const &T = llvm::cast<model::TypedefType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::UnionType: {
+      auto const &T = llvm::cast<model::UnionType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    default:
+      revng_abort();
+    }
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getUnwrappedType(const model::Type &ModelType,
+                   bool RequireComplete = false,
+                   const bool Const = false) {
+    if (const auto *const T = llvm::dyn_cast<model::PrimitiveType>(&ModelType))
+      rc_return getPrimitiveType(*T, Const);
+
+    auto const getDefinedType = [&](const auto Attr) -> clift::ValueType {
+      return make<clift::DefinedType>(Attr, getBool(Const));
+    };
+
+    if (const auto It = Cache.find(ModelType.ID()); It != Cache.end())
+      rc_return getDefinedType(It->second);
+
+    const clift::TypeDefinition Attr = getTypeAttribute(ModelType,
+                                                        RequireComplete);
+
+    if (not Attr)
+      rc_return nullptr;
+
+    if (RequireComplete) {
+      const auto R = Cache.try_emplace(ModelType.ID(), Attr);
+      revng_assert(R.second);
+    }
+
+    rc_return getDefinedType(Attr);
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getQualifiedType(const model::QualifiedType &ModelType,
```

Comment:
`getValue` type I guess. Otherwise a caller expects to get a `QualifiedType`, not a `clift::ValueType`.

Reply:
I renamed it to `getQualifiedValueType`.